Add `entries` and `from_entries`, similar to JavaScript's `Object.entries` and `Object.fromEntries`

someonewithpc opened the pull request:
Previous title: Add `enumerate`, similar to Python's

This PR adds an `enumerate(Transversable|array, int=)` function that acts like [Python's `enumerate`](https://docs.python.org/3.8/library/functions.html#enumerate)

```php
foreach (enumerate($this->hash) as $i => [$k => $v]) {
    // ...
}
```

My use case is to iterate over an array with string keys and have an associated integer that can be used as an ID

lstrojny commented:
I like the idea! I would model it after JavaScripts `Object.fromEntries()` and `Object.entries()` function and have both functions available.

someonewithpc commented:
@lstrojny I'm not too versed in JavaScript, but [from MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/entries#return_value):

> An array of the given object's own enumerable string-keyed property [key, value] pairs

So I don't see the connection. These may also be useful, for sure, but unless I'm not understanding it right, it'd be a completely different pair of functions

lstrojny commented:
Just to clarify: `Object.entries()` does exactly what your (and Python's) `enumerate()` does and `Object.fromEntries()` does the opposite.

So this is how it could look for Functional:

```
Functional\entries(['foo' => 'bar', 'baz' => 'bla']); // [['foo', 'bar'], ['baz', 'bla']]
Functional\from_entries([['foo', 'bar'], ['baz', 'bla']]); // ['foo' => 'bar', 'baz' => 'bla']
```

someonewithpc commented:
Ah, I see! Yes, that would be useful, but there's a subtle difference in the way I made it:

```php
Functional\enumerate(['H' => `'Hydrogen']);` // [0 => ['H' => 'Hydrogen]]
```

Difference being that it keeps the key-value map, meaning it can be reversed with `array_merge(...$enumerated)`.

I'm not entirely sure which is more useful in general. The one I implemented was the one I wanted, but I'll defer to you as to which you prefer/think is more useful

jdreesen commented:
Not sure if this destructuring syntax is legal in PHP, though.

```php
foreach (enumerate($this->hash) as $i => [$k => $v]) {
    // ...
}
```

someonewithpc committed:
Add `entries` and `from_entries`, inspired by JavaScript's `Object.entries` (and Python's `enumerate`) and `Object.from_entries`, respectively

Co-authored-by Lars Strojny <lars@strojny.net>

This commits adds:

 - Functional\entries
 - Functional\from_entries
 - InvalidArgumentException::assertPair
 - EntriesFromEntriesTest
 - Documentation for the added functions

someonewithpc commented:
@lstrojny I updated the code as per your recommendations 

The failing tests seem to be unrelated to my changes (I hope)

lstrojny commented on src/Functional/Entries.php:
```diff
@@ -0,0 +1,36 @@
+use Traversable;
+
+/**
+ * Inspired by JavaScript's `Object.entries`, and Python's `enumerate`,
```
```suggestion
 * Inspired by JavaScript’s `Object.entries`, and Python’s `enumerate`,
```

someonewithpc replied:
Is that U+2019: RIGHT SINGLE QUOTATION MARK? Why that change?

jchase replied:
It's typographically correct.

lstrojny commented on src/Functional/FromEntries.php:
```diff
@@ -0,0 +1,35 @@
+use Traversable;
+
+/**
+ * Inspired by JavaScript's `Object.from_entries`,
```
```suggestion
 * Inspired by JavaScript’s `Object.fromEntries`,
```

lstrojny commented on src/Functional/FromEntries.php:
```diff
@@ -0,0 +1,35 @@
+    InvalidArgumentException::assertCollection($collection, __FUNCTION__, 1);
+
+    $aggregation = [];
+    foreach ($collection as [$key, $value]) {
```
Let's throw an exception if it's not a pair

someonewithpc replied:
You mean if collection doesn't have string keys?

someonewithpc replied:
Right, reading this again I see what you mean. Makes sense, but I'm not sure how to check that

lstrojny replied:
Something like this:
```
foreach ($collection as $enty) {
     InvalidArgumentException::assertPair($entry);
}
```

```
public static function assertPair($pair): void
{
    if (!(is_array($pair) || $pair instanceof ArrayAccess) && !isset($pair[0], $pair[1])) {
          … throw
    }
}
```

Then add a few test cases for the `assertPair()` method

someonewithpc commented on src/Functional/Entries.php:
```diff
@@ -0,0 +1,36 @@
+ * @return array
+ * @no-named-arguments
+ */
+function entries($collection, int $start = 0)
```
`enumerate` supports it, and I was going to use it to have the list start at the ID of the next database row

lstrojny replied:
I am in favor of removing it, if somebody needs to reindex they can do that in a separate call

someonewithpc commented:
I've made the changes you requested, I believe. Apologies for the delay

lstrojny commented on src/Functional/FromEntries.php:
```diff
@@ -0,0 +1,37 @@
+    foreach ($collection as $entry) {
+        InvalidArgumentException::assertPair($entry, __FUNCTION__, 1);
+        [$key, $value] = $entry;
+        $aggregation[$key] = $value;
```
One more thing: we should also check that `$key` is fine to be used as an array key, so checking that it's a string or an int

lstrojny commented on src/Functional/FromEntries.php:
```diff
@@ -0,0 +1,37 @@
+
+/**
+ * Inspired by JavaScript’s `Object.fromEntries`,
+ * convert an array of key-value tuples into a key-value map
```
```suggestion
 * convert an array of key-value pairs into a key-value map
```

lstrojny reviewed (changes requested):
Looking very good already. Few small things, see inline. Also: could you update the docs to include it?

someonewithpc commented:
I believe it's finally done :)

lstrojny commented:
Great, thank you for your patience!

lol768 commented:
@lstrojny Has this been released?
